### ISSUE TITLE
Hosts file not working

### ISSUE THREAD

**Issue by Palleri** [hhoffstaette/dnsdist-utils#1]
Hi, Ive tried to implement the lua script without success.

Ive added these files under a directory
```
hosts.lua
servers.lua
ttl.lua
```
Ive also added these lines in my dnsdist.conf
```
-- Add dnsdist config dir to Lua search path
package.path = package.path .. ";/etc/dnsdist/?.lua"
require "hosts"
addHosts("/etc/hosts")
```

The blockDomain is working, so the lua configuration is working. But the hosts dns does not resolve from the hosts file.

**Comment by hhoffstaette** [comment]
Hmm..that should work since you get no unrelated errors. Maybe your `/etc/hosts` file contains errors or is formatted somehow that trips up the parser; reading the hosts entries is notoriously difficult and this was my first Lua code.
Do you see any entries when you run `showRules()` in the console?

**Comment by Palleri** [comment]
Thank you for answering.
This is my hosts file
```
127.0.0.1 localhost
127.0.1.1 test
10.0.10.1 bla.bla
```
I do not see anything when I execute showRules()
Nor do I see anything in the webui.

Still nothing under rules when I try your example.
```
> addHosts("/etc/dnsdist/test")
[string "return addHosts("/etc/dnsdist/test")"]:1: attempt to call global 'addHosts' (a nil value)
```

**Comment by hhoffstaette** [comment]
> Thank you for answering. This is my hosts file
> 
> ```
> 127.0.0.1 localhost
> 127.0.1.1 test
> 10.0.10.1 bla.bla
> ```

Looks fine and should work.

> ```
> > addHosts("/etc/dnsdist/test")
> [string "return addHosts("/etc/dnsdist/test")"]:1: attempt to call global 'addHosts' (a nil value)
> ```

This means that `hosts.lua` has not been loaded and the function is unknown. Unfortunately that is all I can tell from my end.

**Comment by Palleri** [comment]
I entered the console via this ` dnsdist -l 127.0.0.1:5300 -C /dev/null` and then i read that this start it with the default config.

Now when I entered the console via dnsdist -l 127.0.0.1:5300 I can add the hosts via `addHosts("/etc/hosts")`
But It does not show up in rules :(

**Comment by hhoffstaette** [comment]
Only ideas I have left:
- make sure you have only one instance of the daemon started
- you enter the console to a running dnsdist daemon with `dnsdist -c` (client mode), otherwise you keep starting new fresh daemon instances

**Comment by Palleri** [comment]
Before you wrote back to me I only edited the dnsdist.conf and restarted the service. Didnt really know I could use the console.
Ive restarted the server with the config we mentioned at the top. Cannot see any rules added.

Ive tried to enter the console with -c.
I have no idea if you can do this, but this happened when I tried to add the package.path and require "hosts" this way.

```
root@HAProxy:/etc/dnsdist# dnsdist -c
> package.path = package.path .. ";/etc/dnsdist/lua/?.lua"
> require "hosts"
Command returned an object we can't print: Trying to cast a lua variable from "boolean" to "N5boost8optionalINS_7variantINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEJSt10shared_ptrI15DownstreamStateEP11ClientStateSt13unordered_mapIS7_dSt4hashIS7_ESt8equal_toIS7_ESaISt4pairIKS7_dEEEEEEEE"
> addHosts("/etc/hosts")
> showRules()
#     Matches Rule                                                     Action
0         544 All                                                      log to /var/log/dnsdist.log
```

Once again, thank you for responding :D

**Comment by hhoffstaette** [comment]
This is getting very confusing, please don't do random things in random order. :(

- in the first comment your lua files are in `/etc/dnsdist`, now you add `/etc/dnsdist/lua` to the package path, please pick one. 
- as far as I know the package path must be in the config *before* the server is set up, I don't think it can be set afterwards or from the client.
- I just started my daemon without any rules loaded at start (but the package path set), entered the client, executed `require "hosts"` followed by `addHosts(..)` and the rules showed up. I don't really know what else to do.. ¯\\\_(ツ)_/¯

**Comment by hhoffstaette** [comment]
Also important: you said that the blocklist worked. Adding hosts is basically the same, it just reads & parses a different file. So whatever is going on here is very specific to the hosts file.

**Comment by Palleri** [comment]
The reason for why the path changed is because I changed it. I already have another lua file in that path. So I decided to change it if in some case there is an issue with the import package thingy due to this line. `";/etc/dnsdist/lua/?.lua"`
I suspect that the ? is a wildcard. So I moved your files to the folder lua. My bad I didnt explain that.

If I stop the server with service dnsdist stop.
Then change the config with only the package line as such
```
package.path = package.path .. ";/etc/dnsdist/lua/?.lua"
```
Start the server
```
service dnsdist start
```
Entered the console 
```
root@HAProxy:/etc/dnsdist# dnsdist -c
> require "hosts"
Command returned an object we can't print: Trying to cast a lua variable from "boolean" to "N5boost8optionalINS_7variantINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEJSt10shared_ptrI15DownstreamStateEP11ClientStateSt13unordered_mapIS7_dSt4hashIS7_ESt8equal_toIS7_ESaISt4pairIKS7_dEEEEEEEE"
> addHosts("/etc/hosts")
> showRules()
#     Matches Rule                                                     Action
0          61 All                                                      log to /var/log/dnsdist.log
>
```

Ive tried to restart the whole server and have this config aswell
```
package.path = package.path .. ";/etc/dnsdist/lua/?.lua"
require "hosts"
addHosts("/etc/hosts")
```
Without any luck

**Comment by hhoffstaette** [comment]
Reproducing this in a clean installation (not my production environment, obviously..):

- I created a fresh config with only the lua package statement:
```
$cat /etc/dnsdist/dnsdist.conf

package.path = package.path .. ";/etc/dnsdist/?.lua"

```
- I copied the unmodified `hosts.lua` from the master branch to `/etc/dnsdist`
- I saved your host entries to `/tmp/hosts`:
```
$cat /tmp/hosts 
127.0.0.1 localhost
127.0.1.1 test
10.0.10.1 bla.bla#
```

The first thing I notice is that there's a missing CR at the end of the hosts file, but that does not really matter. Now I start the daemon directly in the foreground:
```
$dnsdist   
dnsdist 1.9.6 comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it according to the terms of the GPL version 2
Raised send buffer to 6291456 for local address '127.0.0.1:53'
Raised receive buffer to 6291456 for local address '127.0.0.1:53'
Listening on 127.0.0.1:53
ACL allowing queries from: 10.0.0.0/8, 100.64.0.0/10, 127.0.0.0/8, 169.254.0.0/16, 172.16.0.0/12, 192.168.0.0/16, ::1/128, fc00::/7, fe80::/10
Console ACL allowing connections from: 127.0.0.0/8, ::1/128
No downstream servers defined: all packets will get dropped
> Polled security status of version 1.9.6 at startup, no known issues reported: OK
```
There are no rules:
```
> showRules()
#   Name                             Matches Rule                Action
```
We import the hosts extension:
```
> require "hosts"
Command returned an object we can't print: Trying to cast a lua variable from "boolean" to "N5boost8optionalINS_7variantINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEJSt10shared_ptrI15DownstreamStateEP11ClientStateSt13unordered_mapIS7_dSt4hashIS7_ESt8equal_toIS7_ESaISt4pairIKS7_dEEEEEEEE"
```
The warning can be safely ignored; it just says that the return value cannot be printed. We could silence this by saying `ok=require(..)`.

Finally we load the specific hosts file:
```
> addHosts("/tmp/hosts")
> showRules()
#   Name                             Matches Rule                Action
0   test                             0 qname in test.            spoof in 127.0.1.1 
1   bla.bla                          0 qname in bla.bla.         spoof in 10.0.10.1 
> 
```
Long story short, this works exactly as advertised: `localhost` is ignored and two rules are created. Please try to reproduce these steps in **exactly** the same way.

Final question: which Lua version is your dnsdist using? If you're not sure you can find out with `ldd`:
```
$ldd /usr/bin/dnsdist | grep lua
	libluajit-5.1.so.2 => /usr/lib64/libluajit-5.1.so.2 (0x00007f9f86aa3000)
```
I have only ever tested with luajit, so there might be an incompatibility with other versions.

**Comment by hhoffstaette** [comment]
@Palleri Did you have a chance to try to reproduce the steps I outlined?
